compiler/src/dmd/statementsem.d

```diff
@@ -3901,21 +3901,24 @@ private extern(D) Expression applyAssocArray(ForeachStatement fs, Expression fld
     if (dim == 2)
     {
         Type ti = (isRef ? taa.index.addMod(MODFlags.const_) : taa.index);
-        if (isRef ? !ti.constConv(ta) : !ti.implicitConvTo(ta))
+        // TODO don't require sizes to match
+        // See https://github.com/dlang/dmd/issues/21456#issuecomment-2981509259
+        if (isRef ? !ti.constConv(ta) : !ti.implicitConvTo(ta) || ti.size != ta.size)
         {
-            error(fs.loc, "`foreach`: index must be type `%s`, not `%s`",
-                     ti.toChars(), ta.toChars());
+            error(fs.loc, "`foreach`: index parameter `%s%s` must be type `%s`, not `%s`",
+                 isRef ? "ref ".ptr : "".ptr, p.toChars(), ti.toChars(), ta.toChars());
             return null;
         }
         p = (*fs.parameters)[1];
         isRef = (p.storageClass & STC.ref_) != 0;
         ta = p.type;
     }
     Type taav = taa.nextOf();
-    if (isRef ? !taav.constConv(ta) : !taav.implicitConvTo(ta))
+    // TODO don't require sizes to match
+    if (isRef ? !taav.constConv(ta) : !taav.implicitConvTo(ta) || taav.size != ta.size)
```

Comment:
I don't think the comparison of the size is correct, the generated lambda should accept the correct parameter types instead. I have patched it here in `foreachBodyToFunction`: https://github.com/dlang/dmd/pull/21066/commits/d2b3b0082a4450a93c5899019e8f00c6a7c66acb#diff-11fb0212c55c05ed6e743e6776abb3c56996381760c8725255c622f96f5d9d4bR4032, but that is part of the template implementation of AAs. Sorry, linking into the diff doesn't seem to work well.

Reply:
OK, so it looks like #21456 is fixed? I can split the error message part out of this PR.

Reply:
It's not merged yet, but I can extract the fix to another PR. This will also cut down on https://github.com/dlang/dmd/pull/21066 a tiny bit.